### ISSUE TITLE
Idea: add index numbers to all messages

### ISSUE THREAD

**Issue by finetuned89** [ocpi/ocpi#156]
There have been some discussions on what to do if a message cannot be delivered to a partner. Since all OCPI messages are an attempt to sync the state of objects between partners, a missed message might cause the state of an object to be false, at least until that object is updated again or until the parter pulls the object from the source system. But with partial object pushes in PATCH requests, a certain property of an object might be false for a long time,

For the receiving partner it is impossible to detect if they have somehow missed a message. Discussions of a meta-channel, some way to re-play messages on demand or other esoteric solutions always concluded this would be too complex.

A simple way to solve this might be to add index numbers to each message for a partner, so they can tell if they have missed a message if an index number is missing. If a hole is detected, the party could pull all objects again from the source system, or we could add a simple meta-module to v2.2 that allows a partner to pull a specific message or range of messages from the source system again.

If you think this is a fruitful and helpful idea, I could write up a more detailed proposal that we could take into the v2.2 discussions next year.

**Comment by poohsen** [comment]
To me this sounds like another reason to consider issue #146. Would your idea still have merit if we dropped PATCH?

**Comment by poohsen** [comment]
so do I understand correctly that each push request would include that sequence number (e.g. in a header) and it would always be incremented by 1 for a new request?

**Comment by finetuned89** [comment]
@poohsen yes that is what I had in mind, include a sequence number in the header that gets incremental for each request. It would still have merit even if we dropped PATCH, because you could still unknowingly miss a POST or PUT request and leave your system in an invalid state (e.g. a token that is not accepted in your system, a location that is no longer available, etc).

**Comment by poohsen** [comment]
I see how this would help with the syncing. But I'm a little reluctant to add mandatory machinery to the protocol. After all, the current counter value needs to be stored for every OCPI module by every implementing party. 
OTOH, the issue that this tries to combat is probably not as visible as it should be for all of us and in the end it's the drivers that suffer from out-of-date data (and our credibility suffers as a consequence). 
Correct me if I'm wrong but this could be an optional feature that the protocol could recommend rather than mandate, right? Then one could try and persuade one's partners to implement it if complaints about stale data are received on a regular basis. And on the client side one is also free to use or ignore that sequence number
